### ISSUE TITLE
Dependency conflicts on org.apache.httpcomponents:httpclient, leading to inconsistent program behaviors

### ISSUE THREAD

**Issue by HelloCoCooo** [freedomotic/freedomotic#490]
Hi, in **freedomotic/plugins/devices/pURL**, there are mulptiple versions of library **org.apache.httpcomponents:httpclient**. However, according to Maven's dependency management strategy: **_"first declaration wins"_**, only **org.apache.httpcomponents:httpclient:4.3.1** can be loaded, and **org.apache.httpcomponents:httpclient:4.5.2** will be shadowed.

As shown in the following figure, your project expects to invoke method **_<org.apache.http.client.config.RequestConfig: toString()Ljava/lang/String;>_**  in library _**org.apache.httpcomponents:httpclient:4.5.2**_ (along the _original dependency path_). As it has been shadowed, this method defined in **_org.apache.httpcomponents:httpclient:4.3.1_** is actually forced to be referenced via the following invocation path (along the _actual dependency path_):

```
<com.freedomotic.plugins.devices.purl.Purl: <init>()V> /home/wwww/sensor/unzip/freedomotic-5.6.0-rc4/plugins/devices/pURL/target/classes
<com.freedomotic.api.Protocol: <init>(Ljava/lang/String;Ljava/lang/String;)V> /home/wwww/.m2/repository/com/freedomotic/freedomotic-core/5.6-SNAPSHOT/freedomotic-core-5.6-SNAPSHOT.jar
<com.freedomotic.api.Plugin: <init>(Ljava/lang/String;Ljava/lang/String;)V> /home/wwww/.m2/repository/com/freedomotic/freedomotic-core/5.6-SNAPSHOT/freedomotic-core-5.6-SNAPSHOT.jar
<com.freedomotic.api.Plugin: init()V> /home/wwww/.m2/repository/com/freedomotic/freedomotic-core/5.6-SNAPSHOT/freedomotic-core-5.6-SNAPSHOT.jar
<com.freedomotic.api.Plugin: setDescription(Ljava/lang/String;)V> /home/wwww/.m2/repository/com/freedomotic/freedomotic-core/5.6-SNAPSHOT/freedomotic-core-5.6-SNAPSHOT.jar
<com.freedomotic.bus.BusServiceImpl: send(Lcom/freedomotic/api/EventTemplate;)V> /home/wwww/.m2/repository/com/freedomotic/freedomotic-core/5.6-SNAPSHOT/freedomotic-core-5.6-SNAPSHOT.jar
<com.freedomotic.events.ProtocolRead: getDefaultDestination()Ljava/lang/String;> /home/wwww/.m2/repository/com/freedomotic/freedomotic-core/5.6-SNAPSHOT/freedomotic-core-5.6-SNAPSHOT.jar
<com.freedomotic.core.TriggerCheck: check(Lcom/freedomotic/api/EventTemplate;Lcom/freedomotic/reactions/Trigger;)Z> /home/wwww/.m2/repository/com/freedomotic/freedomotic-core/5.6-SNAPSHOT/freedomotic-core-5.6-SNAPSHOT.jar
<com.freedomotic.core.TriggerCheck: resolveTrigger(Lcom/freedomotic/api/EventTemplate;Lcom/freedomotic/reactions/Trigger;)Lcom/freedomotic/reactions/Trigger;> /home/wwww/.m2/repository/com/freedomotic/freedomotic-core/5.6-SNAPSHOT/freedomotic-core-5.6-SNAPSHOT.jar
<com.freedomotic.core.Resolver: resolve(Lcom/freedomotic/reactions/Trigger;)Lcom/freedomotic/reactions/Trigger;> /home/wwww/.m2/repository/com/freedomotic/freedomotic-core/5.6-SNAPSHOT/freedomotic-core-5.6-SNAPSHOT.jar
<com.freedomotic.core.Resolver: performSubstitutionInTrigger(Lcom/freedomotic/reactions/Trigger;)V> /home/wwww/.m2/repository/com/freedomotic/freedomotic-core/5.6-SNAPSHOT/freedomotic-core-5.6-SNAPSHOT.jar
<com.freedomotic.core.Resolver: scriptEvaluation(Ljavax/script/ScriptEngine;Ljava/lang/String;Ljava/lang/String;)Ljava/lang/String;> /home/wwww/.m2/repository/com/freedomotic/freedomotic-core/5.6-SNAPSHOT/freedomotic-core-5.6-SNAPSHOT.jar
<org.apache.http.client.config.RequestConfig: toString()Ljava/lang/String;>
```

![freedomotic](https://user-images.githubusercontent.com/46306510/93975731-78224180-fdaa-11ea-8e5b-024b915c27a4.png)

Although both of these conflicting libraries contain the referenced methods (**_with the same signature_**), they have **_different implementations_**. This issue will not cause runtime crashes, but it can introduce inconsistent semantic program hehaviors----

Code snippet of **_<org.apache.http.client.config.RequestConfig: toString()Ljava/lang/String;>_** in _**org.apache.httpcomponents:httpclient:4.5.2**_  (shadowed but expected to invoke method):

<details><summary>detailed method body</summary>

```java
public String toString() {
        final StringBuilder builder = new StringBuilder();
        builder.append("[");
        builder.append("expectContinueEnabled=").append(expectContinueEnabled);
        builder.append(", proxy=").append(proxy);
        builder.append(", localAddress=").append(localAddress);
        builder.append(", cookieSpec=").append(cookieSpec);
        builder.append(", redirectsEnabled=").append(redirectsEnabled);
        builder.append(", relativeRedirectsAllowed=").append(relativeRedirectsAllowed);
        builder.append(", maxRedirects=").append(maxRedirects);
        builder.append(", circularRedirectsAllowed=").append(circularRedirectsAllowed);
        builder.append(", authenticationEnabled=").append(authenticationEnabled);
        builder.append(", targetPreferredAuthSchemes=").append(targetPreferredAuthSchemes);
        builder.append(", proxyPreferredAuthSchemes=").append(proxyPreferredAuthSchemes);
        builder.append(", connectionRequestTimeout=").append(connectionRequestTimeout);
        builder.append(", connectTimeout=").append(connectTimeout);
        builder.append(", socketTimeout=").append(socketTimeout);
        builder.append(", contentCompressionEnabled=").append(contentCompressionEnabled);
        builder.append("]");
        return builder.toString();
    }
```

</details>

Code snippet of **_<org.apache.http.client.config.RequestConfig: toString()Ljava/lang/String;>_** in _**org.apache.httpcomponents:httpclient:4.3.1**_ (loaded version):

<details><summary>detailed method body</summary>

```java
public String toString() {
        final StringBuilder builder = new StringBuilder();
        builder.append(", expectContinueEnabled=").append(expectContinueEnabled);
        builder.append(", proxy=").append(proxy);
        builder.append(", localAddress=").append(localAddress);
        builder.append(", staleConnectionCheckEnabled=").append(staleConnectionCheckEnabled);
        builder.append(", cookieSpec=").append(cookieSpec);
        builder.append(", redirectsEnabled=").append(redirectsEnabled);
        builder.append(", relativeRedirectsAllowed=").append(relativeRedirectsAllowed);
        builder.append(", maxRedirects=").append(maxRedirects);
        builder.append(", circularRedirectsAllowed=").append(circularRedirectsAllowed);
        builder.append(", authenticationEnabled=").append(authenticationEnabled);
        builder.append(", targetPreferredAuthSchemes=").append(targetPreferredAuthSchemes);
        builder.append(", proxyPreferredAuthSchemes=").append(proxyPreferredAuthSchemes);
        builder.append(", connectionRequestTimeout=").append(connectionRequestTimeout);
        builder.append(", connectTimeout=").append(connectTimeout);
        builder.append(", socketTimeout=").append(socketTimeout);
        builder.append("]");
        return builder.toString();
    }
```

</details>

## **Dependency tree--**

[INFO] com.freedomotic:pURL:jar:3.0
[INFO] +- junit:junit:jar:4.11:test
[INFO] |  \- org.hamcrest:hamcrest-core:jar:1.3:test
[INFO] +- com.freedomotic:freedomotic-core:jar:5.6-SNAPSHOT:provided
[INFO] |  +- org.slf4j:jul-to-slf4j:jar:1.7.25:compile (scope managed from provided)
[INFO] |  |  \- org.slf4j:slf4j-api:jar:1.7.25:compile
[INFO] |  +- com.freedomotic:freedomotic-model:jar:5.6-SNAPSHOT:provided
[INFO] |  |  +- (org.slf4j:jul-to-slf4j:jar:1.7.25:compile - scope managed from provided; omitted for duplicate)
[INFO] |  |  +- uk.gov.nationalarchives.thirdparty.netbeans:org-openide-util-lookup:jar:7.2:provided
[INFO] |  |  \- com.fasterxml.jackson.core:jackson-annotations:jar:2.8.9:provided
[INFO] |  +- org.apache.activemq:activemq-all:jar:5.15.8:provided
[INFO] |  +- org.eclipse.jetty.aggregate:jetty-all-server:jar:8.1.18.v20150929:provided
[INFO] |  |  +- org.eclipse.jetty.orbit:javax.servlet:jar:3.0.0.v201112011016:provided
[INFO] |  |  +- org.eclipse.jetty.orbit:javax.security.auth.message:jar:1.0.0.v201108011116:provided
[INFO] |  |  +- org.eclipse.jetty.orbit:javax.mail.glassfish:jar:1.4.1.v201005082020:provided
[INFO] |  |  |  \- (org.eclipse.jetty.orbit:javax.activation:jar:1.1.0.v201105071233:provided - omitted for duplicate)
[INFO] |  |  +- org.eclipse.jetty.orbit:javax.activation:jar:1.1.0.v201105071233:provided
[INFO] |  |  \- org.eclipse.jetty.orbit:javax.annotation:jar:1.1.0.v201108011116:provided
[INFO] |  +- com.thoughtworks.xstream:xstream:jar:1.4.8:provided
[INFO] |  |  +- xmlpull:xmlpull:jar:1.1.3.1:provided
[INFO] |  |  \- xpp3:xpp3_min:jar:1.1.4c:provided
[INFO] |  +- com.google.inject:guice:jar:3.0:provided
[INFO] |  |  +- javax.inject:javax.inject:jar:1:provided
[INFO] |  |  \- aopalliance:aopalliance:jar:1.0:provided
[INFO] |  +- com.google.inject.extensions:guice-assistedinject:jar:3.0:provided
[INFO] |  |  \- (com.google.inject:guice:jar:3.0:provided - omitted for duplicate)
[INFO] |  +- org.apache.shiro:shiro-core:jar:1.4.2:provided
[INFO] |  |  +- org.apache.shiro:shiro-lang:jar:1.4.2:provided
[INFO] |  |  |  \- (org.slf4j:slf4j-api:jar:1.7.25:compile - version managed from 1.7.21; scope managed from provided; omitted for duplicate)
[INFO] |  |  +- org.apache.shiro:shiro-cache:jar:1.4.2:provided
[INFO] |  |  |  \- (org.apache.shiro:shiro-lang:jar:1.4.2:provided - omitted for duplicate)
[INFO] |  |  +- org.apache.shiro:shiro-crypto-hash:jar:1.4.2:provided
[INFO] |  |  |  +- (org.apache.shiro:shiro-lang:jar:1.4.2:provided - omitted for duplicate)
[INFO] |  |  |  \- org.apache.shiro:shiro-crypto-core:jar:1.4.2:provided
[INFO] |  |  |     \- (org.apache.shiro:shiro-lang:jar:1.4.2:provided - omitted for duplicate)
[INFO] |  |  +- org.apache.shiro:shiro-crypto-cipher:jar:1.4.2:provided
[INFO] |  |  |  +- (org.apache.shiro:shiro-lang:jar:1.4.2:provided - omitted for duplicate)
[INFO] |  |  |  \- (org.apache.shiro:shiro-crypto-core:jar:1.4.2:provided - omitted for duplicate)
[INFO] |  |  +- org.apache.shiro:shiro-config-core:jar:1.4.2:provided
[INFO] |  |  |  \- (org.apache.shiro:shiro-lang:jar:1.4.2:provided - omitted for duplicate)
[INFO] |  |  +- org.apache.shiro:shiro-config-ogdl:jar:1.4.2:provided
[INFO] |  |  |  +- (org.apache.shiro:shiro-lang:jar:1.4.2:provided - omitted for duplicate)
[INFO] |  |  |  +- (org.apache.shiro:shiro-config-core:jar:1.4.2:provided - omitted for duplicate)
[INFO] |  |  |  +- (org.apache.shiro:shiro-event:jar:1.4.2:provided - omitted for duplicate)
[INFO] |  |  |  +- commons-beanutils:commons-beanutils:jar:1.9.3:provided
[INFO] |  |  |  |  \- commons-collections:commons-collections:jar:3.2.2:provided
[INFO] |  |  |  \- (org.slf4j:slf4j-api:jar:1.7.25:compile - version managed from 1.7.21; scope managed from provided; omitted for duplicate)
[INFO] |  |  \- org.apache.shiro:shiro-event:jar:1.4.2:provided
[INFO] |  |     \- (org.apache.shiro:shiro-lang:jar:1.4.2:provided - omitted for duplicate)
[INFO] |  +- org.codehaus.jackson:jackson-core-lgpl:jar:1.3.5:provided
[INFO] |  +- org.codehaus.jackson:jackson-mapper-lgpl:jar:1.3.5:provided
[INFO] |  |  \- (org.codehaus.jackson:jackson-core-lgpl:jar:1.3.5:provided - omitted for duplicate)
[INFO] |  +- com.google.code.gson:gson:jar:2.2.4:provided
[INFO] |  +- commons-io:commons-io:jar:2.5:provided
[INFO] |  +- org.apache.commons:commons-lang3:jar:3.6:provided
[INFO] |  +- org.scream3r:jssc:jar:2.8.0:provided
[INFO] |  +- **(org.apache.httpcomponents:httpclient:jar:4.5.2:provided - omitted for conflict with 4.3.1)**
[INFO] |  \- com.google.guava:guava:jar:18.0:provided
[INFO] +- org.apache.httpcomponents:httpcore:jar:4.3:compile
[INFO] \- **org.apache.httpcomponents:httpclient:jar:4.3.1:compile**
[INFO]    +- (org.apache.httpcomponents:httpcore:jar:4.3:compile - omitted for duplicate)
[INFO]    +- commons-logging:commons-logging:jar:1.1.3:compile
[INFO]    \- commons-codec:commons-codec:jar:1.6:compile

## **Suggested solutions:**

**Solution1:** Update direct dependency org.apache.httpcomponents:httpclient from 4.3.1 to 4.5.2.

**Solution2:** Remove the dependency org.apache.httpcomponents:httpclient:4.3.1.

Thanks.
Best regards,
Coco

**Comment by HelloCoCooo** [comment]
Executing the following test case on **org.apache.httpcomponents:httpclient:4.5.2** and **4.3.1** separately, the risky method **_<org.apache.http.client.config.RequestConfig: toString()Ljava/lang/String;>_** will get different return values:

```java
@Test(timeout = 4000)
    public void test01()  throws Throwable  {
        RequestConfig requestConfig0 = RequestConfig.DEFAULT;
        String string0 = requestConfig0.toString();
        assertEquals("[expectContinueEnabled=false, " +
            "proxy=null, " +
            "localAddress=null, " +
            "cookieSpec=null, " +
            "redirectsEnabled=true, " +
            "relativeRedirectsAllowed=true, " +
            "maxRedirects=50, " +
            "circularRedirectsAllowed=false, " +
            "authenticationEnabled=true, " +
            "targetPreferredAuthSchemes=null, " +
            "proxyPreferredAuthSchemes=null, " +
            "connectionRequestTimeout=-1, " +
            "connectTimeout=-1, " +
            "socketTimeout=-1, " +
            "contentCompressionEnabled=true]", string0);
    }
```

Output results:

```java
string0 == ", expectContinueEnabled=false, proxy=null, localAddress=null, staleConnectionCheckEnabled=true, cookieSpec=null, redirectsEnabled=true, relativeRedirectsAllowed=true, maxRedirects=50, circularRedirectsAllowed=false, authenticationEnabled=true, targetPreferredAuthSchemes=null, proxyPreferredAuthSchemes=null, connectionRequestTimeout=-1, connectTimeout=-1, socketTimeout=-1]"   
  //On **org.apache.httpcomponents:httpclient:4.3.1**

string0 == "[expectContinueEnabled=false, proxy=null, localAddress=null, cookieSpec=null, redirectsEnabled=true, relativeRedirectsAllowed=true, maxRedirects=50, circularRedirectsAllowed=false, authenticationEnabled=true, targetPreferredAuthSchemes=null, proxyPreferredAuthSchemes=null, connectionRequestTimeout=-1, connectTimeout=-1, socketTimeout=-1, contentCompressionEnabled=true]"    
  //On **org.apache.httpcomponents:httpclient:4.5.2**
```

**Comment by HelloCoCooo** [comment]
@mcicolella Could please help me check this issue?
May I pull a request to fix it?
Thanks again.

**Comment by mcicolella** [comment]
Hi @HelloCoCooo, thanks for your detailed report.
Maybe the solution 2 could be adopted. Also "push" plugin requires the same dependency but it uses a different version.
Please check if the project compiles and runs correctly and then submit a PR if you want.
Thanks.

**Comment by mcicolella** [comment]
Fixed in #491
